reasoner/planner/ReasonerPlanner.java

```diff
@@ -101,6 +138,56 @@ static Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies(Set
         return deps;
     }
 
+    public Set<CallMode> triggeredCalls(Concludable concludable, Set<Variable> concludableBounds, Optional<Set<ResolvableConjunction>> dependencyFilter) {
+        Set<CallMode> calls = new HashSet<>();
+        for (Map.Entry<Rule, Set<Unifier>> entry : logicMgr.applicableRules(concludable).entrySet()) {
+            ResolvableConjunction ruleConjunction = entry.getKey().condition().conjunction();
+            if (dependencyFilter.isPresent() && !dependencyFilter.get().contains(ruleConjunction)) {
+                continue;
+            }
+            for (Unifier unifier : entry.getValue()) {
+                assert iterate(concludableBounds).allMatch(v -> v.id().isRetrievable());
+                Set<Identifier.Variable.Retrievable> ruleSideIds = iterate(concludableBounds).flatMap(v -> iterate(unifier.mapping().get(v.id().asRetrievable()))).map(Identifier.Variable::asRetrievable).toSet();
+                Set<Variable> ruleSideBounds = iterate(ruleSideIds)
+                        .filter(id -> ruleConjunction.pattern().retrieves().contains(id))
+                        .map(id -> ruleConjunction.pattern().variable(id)).toSet();
+                calls.add(new CallMode(ruleConjunction, ruleSideBounds));
+            }
+        }
+        return calls;
+    }
+
+    static class CallMode {
+        final ResolvableConjunction conjunction;
+        final Set<Variable> bounds;
+        private final int hash;
+
+        CallMode(ResolvableConjunction conjunction, Set<Variable> bounds) {
+            this.conjunction = conjunction;
+            this.bounds = bounds;
+            this.hash = Objects.hash(conjunction, bounds);
+        }
+
+        @Override
+        public int hashCode() {
+            return this.hash;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            CallMode that = (CallMode) o;
+            return conjunction.equals(that.conjunction) &&
+                    bounds.equals(that.bounds);
+        }
+
+        @Override
+        public String toString() {
+            return conjunction + "::{" + iterate(bounds).reduce("", (x, y) -> (x + ", " + y)) + "}";
+        }
+    }
+
     public static class Plan {
         private final List<Resolvable<?>> elementOrder;
```

Comment:
just realised this should be `resolvableOrder` right? why introduce a new term "element"?